bot/exts/moderation/voice_gate.py

```diff
@@ -160,6 +183,46 @@ async def on_message(self, message: discord.Message) -> None:
         with suppress(discord.NotFound):
             await message.delete()
 
+    @Cog.listener()
+    async def on_voice_state_update(
+            self,
+            member: Member,
+            before: discord.VoiceState,
+            after: discord.VoiceState
+    ) -> None:
+        """Pings a user if they've never joined the voice chat before and aren't verified."""
+        if member.bot:
+            log.trace("User is a bot. Ignore.")
+            return
+
+        # member.voice will return None if the user is not in a voice channel
+        if member.voice is None:
+            log.trace("User not in a voice channel. Ignore.")
+            return
+        else:
+            in_cache = await self.redis_cache.get(member.id, None)
+            if in_cache:
+                log.trace("User already in cache. Ignore.")
+                return
+            else:
+                log.trace("User not in cache and is in a voice channel")
+                verified = any(Roles.voice_verified == role.id for role in member.roles)
+                if verified:
+                    log.trace("User is verified, add to the cache and ignore")
+                    # redis cache does not accept None, so False is used to signify no message
+                    await self.redis_cache.set(member.id, False)
+                    return
+
+                log.trace("User is unverified. Send ping.")
+                await self.bot.wait_until_guild_available()
+                voice_verification_channel = self.bot.get_channel(Channels.voice_gate)
+
+                message = await voice_verification_channel.send(f"Hello, {member.mention}! {VOICE_PING}")
+                await self.redis_cache.set(member.id, message.id)
+
+                # Message will try to be deleted after 1 minutes. If it fails, it'll do so silently
+                await message.delete(delay=GateConf.voice_ping_delete_delay)
```

Comment:
Looking good, thanks for addressing all the feedback.

We no longer need many of the else clauses here because of having adjusted the logic to return early. Due to the rewrite, we could have just e.g.:
```python
if member.voice is None:
    ...
    return

in_cache = ...

if in_cache:
    ...
    return

verified = ...
```

The early return helps us avoid the deep indentation, which is difficult to follow.

Reply:
Addressed in latest commit.